File: 6-TypeNode/backend/database/genre-dal.ts
import MysqlPersistence from '../database/mysql-persistence'

type genreDalDTO = {
    id: number | string,
    genre: {
        id: number | string,
        name: string
    }
    bdObj: MysqlPersistence,
    conn: typeof globalThis.connection
}
//! bdOObj param is really necessary ?
const listGenres = async ({ bdObj, conn }: genreDalDTO) => {
    const res = await bdObj.select({
        query:'SELECT gr_id as \'id\', name FROM Genre',
        conn
    })
    return res
}

const readGenre = async ({ id, bdObj, conn }: genreDalDTO): Promise<genreDalDTO["genre"]> => {
    const res = await bdObj.select({
        query: 'SELECT gr_id as \'id\', name FROM Genre where gr_id = ?',
        conn,
        param: [id]
    }) as any
    return res[0]
}

const createGenre = async ({ genre, bdObj, conn }: genreDalDTO) => {
    const res = await bdObj.execute({
        query:'insert into genre (name) values (?)',
        conn,
        param: [genre.name]
    })
    genre.id = res
    return genre
}

const updateGenre = async ({ genre, bdObj, conn }: genreDalDTO) => {
    const res = await bdObj.execute({
        query: 'update genre set name = ? where gr_id = ?',
        conn,
        param: [genre.name, genre.id]
    })
    return res
}

const delGenre = async ({ id, bdObj, conn }: genreDalDTO) => {
    const res = await bdObj.execute({
        query: 'delete from genre where gr_id = ?',
        conn,
        param: [id]
    })
    return res
}

module.exports = {
    listGenres,
    readGenre,
    createGenre,
    updateGenre,
    delGenre
}

File: 9-Golang/backend/database/genre-dal.go
package database

type Genre struct {
	Id   int    `json:"id"`
	Name string `json:"name"`
}

func GetGenres(conn *Connector) ([]Genre, error) {
	genres := []Genre{}
	rows, err := conn.db.Query("SELECT * FROM Genre")
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	for rows.Next() {
		var genre Genre
		err = rows.Scan(&genre.Id, &genre.Name)
		if err != nil {
			return nil, err
		}
		genres = append(genres, genre)
	}
	return genres, nil
}

func GetGenreById(conn *Connector, id int) (Genre, error) {
	var genre Genre

	err := conn.db.QueryRow("SELECT * FROM Genre WHERE id = ?", id).Scan(&genre.Id, &genre.Name)
	if err != nil {
		return genre, err
	}

	return genre, nil
}

func CreateGenre(conn *Connector, genre *Genre) error {

	_, err := conn.db.Exec("INSERT INTO Genre (name) VALUES (?)", genre.Name)
	if err != nil {
		return err
	}

	return nil
}

func UpdateGenre(conn *Connector, genre *Genre) error {
	_, err := conn.db.Exec("UPDATE Genre SET name = ? WHERE id = ?", genre.Name, genre.Id)
	if err != nil {
		return err
	}

	return nil
}

func DeleteGenre(conn *Connector, id int) error {
	_, err := conn.db.Exec("DELETE FROM Genre WHERE id = ?", id)
	if err != nil {
		return err
	}
	return nil
}
